[DebugInfo] Fix debug info round tripping of types inside functions

augusto2112 opened the pull request:
Types with @_originallyDefinedIn cannot be round tripped,

Types declared inside functions have their mangling affected by the function signature. If the generic signature mentions an @_originallyDefinedIn type, the type inside the function cannot be round tripped either.

This commit disables round tripping for this scenario.

<!--
If this pull request is targeting a release branch, please fill out the
following form:
https://github.com/swiftlang/.github/blob/main/PULL_REQUEST_TEMPLATE/release.md?plain=1

Otherwise, replace this comment with a description of your changes and
rationale. Provide links to external references/discussions if appropriate.
If this pull request resolves any GitHub issues, link them like so:

  Resolves <link to issue>, resolves <link to another issue>.

For more information about linking a pull request to an issue, see:
https://docs.github.com/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue
-->

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/main/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

augusto2112 committed:
[DebugInfo] Fix debug info round tripping of types inside functions

Types with @_originallyDefinedIn cannot be round tripped,

Types declared inside functions have their mangling affected by the
function signature. If the generic signature mentions an
@_originallyDefinedIn type, the type inside the function cannot be round
tripped either.

This commit disables round tripping for this scenario.

augusto2112 commented:
@swift-ci test

hjyamauchi commented:
@augusto2112 It seems like I still locally get a crash with this PR. I will double-check.
```
Failed to reconstruct type for $s11SwiftFormat24DocumentationCommentTextV13extractedFromACSg08CompilerA6Syntax6TriviaV_tcfc4LineL_VD
Original type:
(struct_type decl="SwiftFormat.(file).DocumentationCommentText.init(extractedFrom:).Line@S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\Core\\DocumentationCommentText.swift:56:12")
Please submit a bug report (https://swift.org/contributing/#reporting-bugs) and include the crash backtrace.
Pass '-Xfrontend -disable-round-trip-debug-types' to disable this assertion.
Please submit a bug report (https://swift.org/contributing/#reporting-bugs) and include the crash backtrace.
Stack dump:
0.      Program arguments: C:\\Users\\hiroshi\\source\\b\\5\\bin\\swiftc.exe -frontend -c S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\API\\Configuration+Default.swift S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\API\\Configuration.swift S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\API\\DebugOptions.swift S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\API\\Finding.swift S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\API\\FindingCategorizing.swift S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\API\\Indent.swift S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\API\\Selection.swift S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\API\\SwiftFormatError.swift S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\API\\SwiftFormatter.swift S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\API\\SwiftLinter.swift S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\Core\\Context.swift S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\Core\\DocumentationComment.swift -primary-file S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\Core\\DocumentationCommentText.swift S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\Core\\Finding+Convenience.swift S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\Core\\FindingEmitter.swift S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\Core\\FormatPipeline.swift S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\Core\\FunctionDeclSyntax+Convenience.swift S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\Core\\ImportsXCTestVisitor.swift S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\Core\\LazySplitSequence.swift S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\Core\\LintPipeline.swift S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\Core\\ModifierListSyntax+Convenience.swift S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\Core\\Parsing.swift S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\Core\\Pipelines+Generated.swift S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\Core\\RememberingIterator.swift S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\Core\\Rule.swift S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\Core\\RuleBasedFindingCategory.swift S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\Core\\RuleMask.swift S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\Core\\RuleNameCache+Generated.swift S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\Core\\RuleRegistry+Generated.swift S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\Core\\RuleState.swift S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\Core\\SyntaxFormatRule.swift S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\Core\\SyntaxLintRule.swift S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\Core\\SyntaxProtocol+Convenience.swift S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\Core\\Trivia+Convenience.swift S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\Core\\WithAttributesSyntax+Convenience.swift S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\Core\\WithSemicolonSyntax.swift S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\PrettyPrint\\Comment.swift S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\PrettyPrint\\Indent+Length.swift S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\PrettyPrint\\PrettyPrint.swift S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\PrettyPrint\\PrettyPrintBuffer.swift S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\PrettyPrint\\PrettyPrintFindingCategory.swift S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\PrettyPrint\\Token.swift S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\PrettyPrint\\TokenStreamCreator.swift S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\PrettyPrint\\Verbatim.swift S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\PrettyPrint\\WhitespaceFindingCategory.swift S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\PrettyPrint\\WhitespaceLinter.swift S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\Rules\\AllPublicDeclarationsHaveDocumentation.swift S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\Rules\\AlwaysUseLiteralForEmptyCollectionInit.swift S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\Rules\\AlwaysUseLowerCamelCase.swift S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\Rules\\AmbiguousTrailingClosureOverload.swift S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\Rules\\AvoidRetroactiveConformances.swift S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\Rules\\BeginDocumentationCommentWithOneLineSummary.swift S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\Rules\\DoNotUseSemicolons.swift S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\Rules\\DontRepeatTypeInStaticProperties.swift S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\Rules\\FileScopedDeclarationPrivacy.swift S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\Rules\\FullyIndirectEnum.swift S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\Rules\\GroupNumericLiterals.swift S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\Rules\\IdentifiersMustBeASCII.swift S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\Rules\\NeverForceUnwrap.swift S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\Rules\\NeverUseForceTry.swift S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\Rules\\NeverUseImplicitlyUnwrappedOptionals.swift S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\Rules\\NoAccessLevelOnExtensionDeclaration.swift S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\Rules\\NoAssignmentInExpressions.swift S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\Rules\\NoBlockComments.swift S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\Rules\\NoCasesWithOnlyFallthrough.swift S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\Rules\\NoEmptyLineOpeningClosingBraces.swift S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\Rules\\NoEmptyTrailingClosureParentheses.swift S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\Rules\\NoLabelsInCasePatterns.swift S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\Rules\\NoLeadingUnderscores.swift S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\Rules\\NoParensAroundConditions.swift S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\Rules\\NoPlaygroundLiterals.swift S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\Rules\\NoVoidReturnOnFunctionSignature.swift S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\Rules\\OmitExplicitReturns.swift S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\Rules\\OneCasePerLine.swift S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\Rules\\OneVariableDeclarationPerLine.swift S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\Rules\\OnlyOneTrailingClosureArgument.swift S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\Rules\\OrderedImports.swift S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\Rules\\ReplaceForEachWithForLoop.swift S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\Rules\\ReturnVoidInsteadOfEmptyTuple.swift S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\Rules\\TypeNamesShouldBeCapitalized.swift S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\Rules\\UseEarlyExits.swift S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\Rules\\UseExplicitNilCheckInConditions.swift S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\Rules\\UseLetInEveryBoundCaseVariable.swift S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\Rules\\UseShorthandTypeNames.swift S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\Rules\\UseSingleLinePropertyGetter.swift S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\Rules\\UseSynthesizedInitializer.swift S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\Rules\\UseTripleSlashForDocumentationComments.swift S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\Rules\\UseWhereClausesInForLoops.swift S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\Rules\\ValidateDocumentationComments.swift S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\Utilities\\FileIterator.swift S:\\SourceCache\\swift-format\\Sources\\SwiftFormat\\Utilities\\URL+isRoot.swift -emit-module-path Sources\\SwiftFormat\\CMakeFiles\\SwiftFormat.dir\\Core\\DocumentationCommentText.swift.swiftmodule -emit-module-doc-path Sources\\SwiftFormat\\CMakeFiles\\SwiftFormat.dir\\Core\\DocumentationCommentText.swift.swiftdoc -emit-module-source-info-path Sources\\SwiftFormat\\CMakeFiles\\SwiftFormat.dir\\Core\\DocumentationCommentText.swift.swiftsourceinfo -emit-dependencies-path Sources\\SwiftFormat\\CMakeFiles\\SwiftFormat.dir\\Core\\DocumentationCommentText.swift.obj.d -target x86_64-unknown-windows-msvc -disable-objc-interop -sdk "S:/Program Files/Swift/Platforms/Windows.platform/Developer/SDKs/Windows.sdk" -I S:\\b\\23\\swift -I S:\\SourceCache\\swift-markdown\\Sources\\CAtomic\\include -I S:\\b\\5\\.\\lib\\swift\\host -I "S:\\Program Files\\Swift\\Toolchains\\0.0.0+Asserts\\usr\\include\\cmark_gfm" -I "S:\\Program Files\\Swift\\Toolchains\\0.0.0+Asserts\\usr\\include\\cmark_gfm_extensions" -g -module-link-name SwiftFormat -O -D SwiftFormat_EXPORTS -in-process-plugin-server-path C:\\Users\\hiroshi\\source\\b\\5\\bin\\SwiftInProcPluginServer.dll -plugin-path C:\\Users\\hiroshi\\source\\b\\5\\bin -autolink-library oldnames -autolink-library msvcrt -Xcc -D_MT -Xcc -D_DLL -parse-as-library -module-name SwiftFormat -num-threads 32 -o Sources\\SwiftFormat\\CMakeFiles\\SwiftFormat.dir\\Core\\DocumentationCommentText.swift.obj
1.      Swift version 6.2-dev (LLVM aed73c6d65289c8, Swift ae88aaca8f76c02)
2.      Compiling with effective version 5.10
3.      While evaluating request IRGenRequest(IR Generation for file "S:\SourceCache\swift-format\Sources\SwiftFormat\Core\DocumentationCommentText.swift")
4.      While emitting IR SIL function "@$s11SwiftFormat24DocumentationCommentTextV13extractedFromACSg08CompilerA6Syntax6TriviaV_tcfC".
 for 'init(extractedFrom:)' (at S:\SourceCache\swift-format\Sources\SwiftFormat\Core\DocumentationCommentText.swift:54:10)
```

augusto2112 commented:
@swift-ci test

augusto2112 commented:
@swift-ci test

augusto2112 commented:
@hjyamauchi thanks for testing, could you try again with the latest commit?

adrian-prantl commented on lib/IRGen/IRGenDebugInfo.cpp:
```diff
@@ -2343,11 +2346,16 @@ createSpecializedStructOrClassType(NominalOrBoundGenericNominalType *Type,
-      if (auto *TAT = llvm::dyn_cast<TypeAliasType>(T)) {
-        auto D = TAT->getDecl()->getDeclContext();
+      if (D) {
         if (auto AFD = llvm::dyn_cast<AbstractFunctionDecl>(D)) {
```
```suggestion
        if (auto AFD = llvm::dyn_cast_or_null<AbstractFunctionDecl>(D)) {
```

augusto2112 commented:
@swift-ci smoke test

hjyamauchi commented:
> @hjyamauchi thanks for testing, could you try again with the latest commit?

@augusto2112 Will do. I also added a reduced test here: https://github.com/swiftlang/swift/issues/78326#issue-2753360747

hjyamauchi commented:
> > @hjyamauchi thanks for testing, could you try again with the latest commit?
> 
> @augusto2112 Will do. I also added a reduced test here: [#78326 (comment)](https://github.com/swiftlang/swift/issues/78326#issue-2753360747)

@augusto2112 The updated PR doesn't seem to fix the issue for me.

augusto2112 commented:
That's surprising, let me check again if I'm missing some other case...

hjyamauchi commented:
@augusto2112 It might be useful to use the reduced test from here: https://github.com/swiftlang/swift/issues/78326#issue-2753360747

augusto2112 commented:
@hjyamauchi thanks, I was able to use your reproducer. It actually was not related to @_originallyDefinedIn as I thought, but something else entirely. I'm running the test suite locally and hopefully this will be the last iteration of the fix.

augusto2112 committed:
Remove hardcode to emit types in CompilerSwiftSyntax with ABI mod name

Now that mangling for the debugger respects originally defined in, this
is not needed anymore.

augusto2112 commented:
@swift-ci test

augusto2112 commented:
@hjyamauchi if you're around this should work with your test case now :)

augusto2112 commented:
@swift-ci smoke test

hjyamauchi commented:
> @hjyamauchi if you're around this should work with your test case now :)

Testing

hjyamauchi commented:
It looks good!
